dropshot/src/extractor/body.rs

```diff
@@ -184,38 +189,246 @@ impl UntypedBody {
 impl ExclusiveExtractor for UntypedBody {
     async fn from_request<Context: ServerContext>(
         rqctx: &RequestContext<Context>,
-        mut request: hyper::Request<hyper::Body>,
+        request: hyper::Request<hyper::Body>,
     ) -> Result<UntypedBody, HttpError> {
         let server = &rqctx.server;
-        let body_bytes = http_read_body(
-            request.body_mut(),
-            server.config.request_body_max_bytes,
-        )
-        .await?;
-        Ok(UntypedBody { content: body_bytes })
+        let body = request.into_body();
+        let body_bytes =
+            StreamingBody::new(body, server.config.request_body_max_bytes)
+                .into_bytes_mut()
+                .await?;
+        Ok(UntypedBody { content: body_bytes.freeze() })
     }
 
     fn metadata(
         _content_type: ApiEndpointBodyContentType,
     ) -> ExtractorMetadata {
-        ExtractorMetadata {
-            parameters: vec![ApiEndpointParameter::new_body(
-                ApiEndpointBodyContentType::Bytes,
-                true,
-                ApiSchemaGenerator::Static {
-                    schema: Box::new(
-                        SchemaObject {
-                            instance_type: Some(InstanceType::String.into()),
-                            format: Some(String::from("binary")),
-                            ..Default::default()
-                        }
-                        .into(),
-                    ),
-                    dependencies: indexmap::IndexMap::default(),
-                },
-                vec![],
-            )],
-            extension_mode: ExtensionMode::None,
+        untyped_metadata()
+    }
+}
+
+// StreamingBody: body extractor that provides a streaming representation of the body.
+
+/// An extractor for streaming the contents of the HTTP request body, making the
+/// raw bytes available to the consumer.
+#[derive(Debug)]
+pub struct StreamingBody {
+    body: hyper::Body,
+    cap: usize,
+}
+
+impl StreamingBody {
+    fn new(body: hyper::Body, cap: usize) -> Self {
+        Self { body, cap }
+    }
+
+    /// Not part of the public API. Used only for doctests.
+    #[doc(hidden)]
+    pub fn __from_bytes(data: Bytes) -> Self {
+        let cap = data.len();
+        let stream = futures::stream::iter([Ok::<_, Infallible>(data)]);
+        let body = hyper::Body::wrap_stream(stream);
+        Self { body, cap }
+    }
+
+    /// Converts `self` into a [`BytesMut`], buffering the entire response in memory.
+    ///
+    /// If payloads are expected to be large, consider using [`Self::into_stream`] to
+    /// avoid buffering in memory if possible.
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`HttpError`] if any of the following cases occur:
+    ///
+    /// * A network error occurred.
+    /// * `request_body_max_bytes` was exceeded for this request.
+    pub async fn into_bytes_mut(self) -> Result<BytesMut, HttpError> {
```

Comment:
What do you think about not exposing this publicly?  For the reason you mentioned it's a (small) footgun, and if you want that you can use `UntypedBody` already, right?

Reply:
Yeah, on second thought I think what we should do is recommend you use `UntypedBody`, and let you convert *that* into a `BytesMut`. Can do that in a followup.